rust/src/nasl/builtin/network/socket.rs

```diff
@@ -769,11 +561,160 @@ impl NaslSockets {
         let port = verify_port(port)?;
         let addr = ipstr2ipaddr(context.target())?;
 
-        let socket = Self::open_udp(addr, port)?;
+        let socket = NaslSocket::Udp(UdpConnection::new(addr, port)?);
         let fd = self.add(socket);
 
         Ok(NaslValue::Number(fd as i64))
     }
+
+    /// Get the source port of a open socket
+    #[nasl_function]
+    fn get_source_port(&self, socket: usize) -> Result<NaslValue, FunctionErrorKind> {
+        let handles = self.handles.read().unwrap();
+        let socket = handles
+            .handles
+            .get(socket)
+            .ok_or(FunctionErrorKind::WrongArgument(
+                "the given socket FD does not exist".to_string(),
+            ))?;
+        let port = match socket {
+            NaslSocket::Tcp(conn) => conn.local_addr()?.port(),
+            NaslSocket::Udp(conn) => conn.local_addr()?.port(),
+            NaslSocket::Close => {
+                return Err(FunctionErrorKind::WrongArgument(
+                    "the given socket FD is already closed".to_string(),
+                ))
+            }
+        };
+        Ok(NaslValue::Number(port as i64))
+    }
+
+    /// *any* **ftp_log_in**(user: *string*, pass: *string*, socket: *int*);
+
+    /// **ftp_log_in** takes three named arguments:
+    /// - user: is the user name (it has no default value like “anonymous” or “ftp”)
+    /// - pass: is the password (again, no default value like the user e-mail address)
+    /// - socket: an open socket.
+    #[nasl_function(named(user, pass, socket))]
+    fn ftp_log_in(&self, user: &str, pass: &str, socket: usize) -> Result<bool, FunctionErrorKind> {
+        let mut buf = String::new();
+        match self
+            .handles
+            .write()
+            .unwrap()
+            .handles
+            .get_mut(socket)
+            .ok_or(FunctionErrorKind::WrongArgument(format!(
+                "the given socket FD {socket} does not exist"
+            )))? {
+            NaslSocket::Tcp(conn) => {
+                let n = conn.read_line(&mut buf)?;
+                if n == 0 {
+                    return Err(FunctionErrorKind::Diagnostic(
+                        "FTP server did not respond".to_string(),
+                        None,
+                    ));
+                }
+
+                // Check if buf contains 220
+                if !buf.contains("220") {
+                    return Err(FunctionErrorKind::Diagnostic(
+                        "FTP server did not respond with 220".to_string(),
+                        None,
+                    ));
+                }
+
+                for i in 0..1024 {
+                    let n = conn.read_line(&mut buf)?;
+                    if n == 0 {
+                        return Err(FunctionErrorKind::Diagnostic(
+                            "FTP server did not respond with 220".to_string(),
```

Comment:
Is this a correct error message here? It seems that this was appropriate above but shouldn't this be "FTP server did not respond"? Disclaimer: I don't understand the surrounding code at all since I don't know anything about FTP.